### PR TITLE
Add a new target "ninja clippy" for Rust projects

### PR DESCRIPTION
**Pull request by bonzini**
Running clippy by setting a custom rustc compiler that points to "clippy-driver" is suboptimal, because it requires reconfiguration. Similar to the "ninja scan-build" target for C, add a clippy internal tool that runs clippy-driver on all crates in the project.

This also provide infrastructure to find sibling tools to rustc, such as "clippy-driver" itself or rustdoc, when using the "rustup run TOOLCHAIN rustc" to invoke rustc. Adding rustfmt for example is trivial.

A few extra new features are included as a bonus :angel: 
* `MESON_NUM_PROCESSES` respected every time Meson invokes multiple processes in parallel (supersedes `MESON_TESTTHREADS`
* clang-tidy targets respect `b_colorout`
* clang-tidy target do not mix output from multiple files
* target introspection gains a `machine` entry for each sources group, allowing cross lookup from targets to compilers

### CODE REVIEW

File: mesonbuild/scripts/run_tool.py [hunk 1]
```diff
@@ -3,17 +3,85 @@
...
+    except asyncio.CancelledError:
+        if p:
+            p.kill()
+            await p.wait()
+            return p.returncode or 1
+        else:
+            return 0
+
+async def _run_workers(infos: T.Iterable[Info],
+                       fn: T.Callable[[Info], T.Iterable[T.Coroutine[T.Any, T.Any, int]]]) -> int:
```
Comment (dcbaker) [inline mesonbuild/scripts/run_tool.py:53]: I think you can simplify this to `fn: T.Callable[[Info], T.Iterable[T.Awaitable[int]]]`

Bonus points that it remove the need for `Any`

File: mesonbuild/scripts/run_tool.py [hunk 2]
```diff
@@ -3,17 +3,85 @@
...
+    futures: T.List[asyncio.Future[int]] = []
+    semaphore = asyncio.Semaphore(determine_worker_count())
+
+    async def run_one(worker_coro: T.Coroutine[T.Any, T.Any, int]) -> int:
```
Comment (dcbaker) [inline mesonbuild/scripts/run_tool.py:57]: Same here on Coroutine -> Awaitable.

File: mesonbuild/scripts/run_tool.py [hunk 3]
```diff
@@ -44,29 +112,18 @@ def run_clang_tool(name: str, srcdir: Path, builddir: Path, fn: T.Callable[...,
     suffixes = set(lang_suffixes['c']).union(set(lang_suffixes['cpp']))
     suffixes.add('h')
     suffixes = {f'.{s}' for s in suffixes}
-    futures = []
-    returncode = 0
-    e = concurrent.futures.ThreadPoolExecutor()
-    try:
-        for f in itertools.chain(*globs):
+    for g in globs:
+        for f in g:
```
Comment (dcbaker) [inline mesonbuild/scripts/run_tool.py:118]: Why remove the `itertoos.chain`? Although this would be nicer as `itertools.chain.from(globs)`

File: docs/markdown/IDE-integration.md [hunk 4]
```diff
@@ -93,6 +93,7 @@ can provide code completion for all source files.
 ```json
 {
     "language": "language ID",
+    "machine": "build|host",
```
Comment (dcbaker) [inline docs/markdown/IDE-integration.md:96]: Shouldn't the be `"build|host|unknown"` given the snippet>

Although I wonder if it makes more sense to be `"build" | "host" | null`, since the place where it would be unknown is for custom_target, right?

File: mesonbuild/scripts/clippy.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
+# SPDX-License-Identifier: Apache-2.0
+# Copyright 2024 The Meson development team
+
+from collections import defaultdict
+
```
Comment (dcbaker) [inline mesonbuild/scripts/clippy.py:5]: Delete this newline

File: mesonbuild/scripts/clippy.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
+# SPDX-License-Identifier: Apache-2.0
+# Copyright 2024 The Meson development team
+
```
Comment (dcbaker) [inline mesonbuild/scripts/clippy.py:3]: `from __future__ import annotations`

File: mesonbuild/compilers/detect.py [hunk 7]
```diff
@@ -1042,6 +1042,10 @@ def detect_rust_compiler(env: 'Environment', for_machine: MachineChoice) -> Rust
             version = search_version(out)
 
             cls = rust.ClippyRustCompiler
+            mlog.warning(
```
Comment (dcbaker) [inline mesonbuild/compilers/detect.py:1045]: As the author of the clippy-driver support, please switch to using `FeatureDeprecated` for this so we can remove it in the future.

**Review by dcbaker — CHANGES_REQUESTED** [review]
Generally looks good. Thanks for working on this.

File: mesonbuild/scripts/run_tool.py (hunk 1 above)
Comment (bonzini) [reply mesonbuild/scripts/run_tool.py:53]: Unfortunately you need Coroutine to be able to throw a CancellableError into it. Without the `throw`, you get a "coroutine was never awaited" error.

File: mesonbuild/scripts/run_tool.py [hunk 8]
```diff
@@ -3,17 +3,85 @@
...
+        print(mlog.blue('>>>'), quoted_cmdline, file=sys.stderr)
+        print(mlog.red('not found:'), e.filename, file=sys.stderr)
+        return 1
+    except asyncio.CancelledError:
+        if p:
+            p.kill()
+            await p.wait()
+            return p.returncode or 1
+        else:
+            return 0
```
[the remark this answers is not in the document]
Comment (bonzini) [reply mesonbuild/scripts/run_tool.py:50]: There's a `return p.returncode` above at the end of the `try:` part, I will clarify by pulling it down.

File: docs/markdown/IDE-integration.md (hunk 4 above)
Comment (bonzini) [reply docs/markdown/IDE-integration.md:96]: Right, null or absent makes more sense since (unlike language) the machine is defined even in the introspect-from-meson.build case.

File: mesonbuild/compilers/detect.py (hunk 7 above)
Comment (bonzini) [reply mesonbuild/compilers/detect.py:1045]: This is a user feature not a developer one, so does it make sense to key it on `meson_version`? Anyhow I will look for a similar case and sort it out. :)

Comment (eli-schwartz) [reply mesonbuild/compilers/detect.py:1045]: Indeed, it doesn't make sense to key it on meson_version because it cannot be set from project files. At best, you could set it from a native file.

I am opposed to using FeatureDeprecated for this.

Comment (dcbaker) [reply mesonbuild/compilers/detect.py:1045]: I thinks it's appropriate since clippy-driver is discouraged for use by upstream and we now provide a better option that is supported by upstream.

@eli-schwartz, @jpakkane, @xclaesse: opinions on this?

File: docs/markdown/IDE-integration.md (hunk 4 above)
Comment (dcbaker) [reply docs/markdown/IDE-integration.md:96]: I think I have a slight preference for null, since technically some custom_targets are wrappers around configure/make and thus could have a machine?, but absent would be fine

File: mesonbuild/scripts/run_tool.py (hunk 1 above)
Comment (dcbaker) [reply mesonbuild/scripts/run_tool.py:53]: In that case I think it should be `T.Coroutine[None, None, int]`, since the documentation says the annotations are the same as `Generator`; https://docs.python.org/3/library/typing.html#annotating-generators-and-coroutines

File: mesonbuild/scripts/run_tool.py [hunk 9]
```diff
@@ -3,17 +3,85 @@
...
+    try:
+        p = await asyncio.create_subprocess_exec(*cmdlist,
+                                                 stdin=asyncio.subprocess.DEVNULL,
+                                                 stdout=asyncio.subprocess.PIPE,
+                                                 stderr=asyncio.subprocess.STDOUT)
+        stdo, _ = await p.communicate()
+        if stdo:
+            print(mlog.blue('>>>'), quoted_cmdline)
+            sys.stdout.flush()
+            sys.stdout.buffer.write(stdo)
```
[the remark this answers is not in the document]
Comment (dcbaker) [reply mesonbuild/scripts/run_tool.py:38]: okay, that makes sense.

File: mesonbuild/compilers/detect.py (hunk 7 above)
Comment (eli-schwartz) [reply mesonbuild/compilers/detect.py:1045]: @dcbaker the point is that it's already a `WARNING: ...`, due to using `mlog.warning()`. The only difference that using FeatureDeprecated provides is making the `WARNING: ...` conditional on which minimum version the meson.build file requires due to the functions/syntax used in meson.build.

This is totally unrelated to "can we remove it in the future", which has never been the point of FeatureDeprecated.

Just change:
> `mlog.warning('clippy-driver is not intended as a general purpose compiler.  You can use "ninja clippy" in order to run clippy on a meson project.')`

to


> `mlog.warning('clippy-driver is not intended as a general purpose compiler.  You can use "ninja clippy" in order to run clippy on a meson project. Usee as a general purpose compiler will be removed in a future version of meson')`

and also frankly I don't see this as being covered by meson's stability guarantees. Projects which were written once, continue to work after upgrading meson.

File: docs/markdown/IDE-integration.md (hunk 4 above)
Comment (bonzini) [reply docs/markdown/IDE-integration.md:96]: Right, but I was going for absent exactly because of cases where you do have a machine but Meson cannot place it in the introspection data. Absent leaves the door open for future extension. The "new in 1.7.0" is also missing here in the docs.

**Review by dcbaker — APPROVED** [review]
Thanks, this looks good to me.

**Comment by bonzini** [comment]
Hmm, it's a problem with TypedDict... Looks like I will have to make machine `null`.

**Comment by dcbaker** [comment]
@bonzini you can using `typing_extensions.NotRequired[]`

**Review by dcbaker — APPROVED** [review]

**Comment by eli-schwartz** [comment]
Restarting a potentially flaky job...

File: mesonbuild/scripts/run_tool.py [hunk 10]
```diff
@@ -3,17 +3,86 @@
 
 from __future__ import annotations
 
-import itertools
+import asyncio
+import asyncio.subprocess
```
Comment (eli-schwartz) [inline mesonbuild/scripts/run_tool.py:7]: You don't need to import both of these -- importing a submodule implies importing the top-level module too. :)

File: mesonbuild/scripts/run_tool.py [hunk 11]
```diff
@@ -3,17 +3,86 @@
...
 
-if T.TYPE_CHECKING:
-    import subprocess
+Info = T.TypeVar("Info")
+
+async def run_with_buffered_output(cmdlist: T.List[str]) -> int:
+    """Run the command in cmdlist, buffering the output so that it is
+       not mixed for multiple child processes.  Kill the child on
+       cancellation."""
+    quoted_cmdline = ' '.join(sh_quote(x) for x in cmdlist)
```
Comment (eli-schwartz) [inline mesonbuild/scripts/run_tool.py:28]: What is quoted_cmdline, and why does mtest have a function that isn't mesonlib.join_args for this?

File: mesonbuild/scripts/run_tool.py [hunk 12]
```diff
@@ -3,17 +3,86 @@
...
+    p: T.Optional[asyncio.subprocess.Process] = None
+    try:
+        p = await asyncio.create_subprocess_exec(*cmdlist,
+                                                 stdin=asyncio.subprocess.DEVNULL,
+                                                 stdout=asyncio.subprocess.PIPE,
+                                                 stderr=asyncio.subprocess.STDOUT)
```
Comment (eli-schwartz) [inline mesonbuild/scripts/run_tool.py:34]: I think that this means we kill colored output since tools being run are not going to see isatty() return true anymore... this is a bit unfortunate.

File: mesonbuild/scripts/run_tool.py [hunk 13]
```diff
@@ -3,17 +3,86 @@
...
+
+    if stdo:
+        print(mlog.blue('>>>'), quoted_cmdline, flush=True)
+        sys.stdout.buffer.write(stdo)
+    return p.returncode
+
+async def _run_workers(infos: T.Iterable[Info],
+                       fn: T.Callable[[Info], T.Iterable[T.Coroutine[None, None, int]]]) -> int:
+    futures: T.List[asyncio.Future[int]] = []
+    semaphore = asyncio.Semaphore(determine_worker_count())
```
Comment (eli-schwartz) [inline mesonbuild/scripts/run_tool.py:56]: determine_worker_count() is highly specific to `meson test`. It is in fact identical to multiprocessing.cpu_count() except when the environment variable `$MESON_TESTTHREADS` is exported, and it is obviously incorrect behavior for `ninja clang-format` to be controlled by a "test threads" setting.

File: mesonbuild/scripts/run_tool.py [hunk 14]
```diff
@@ -3,17 +3,86 @@
...
+        loop.add_signal_handler(signal.SIGINT, sigterm_handler)
+        loop.add_signal_handler(signal.SIGTERM, sigterm_handler)
+
+    for i in infos:
+        futures.extend((asyncio.ensure_future(run_one(x)) for x in fn(i)))
+    if not futures:
+        return 0
+
+    try:
+        await complete_all(futures)
```
Comment (eli-schwartz) [inline mesonbuild/scripts/run_tool.py:81]: ... which leaves us with this as the one obviously useful import from mtest.py, which should potentially be refactored instead?

File: mesonbuild/utils/universal.py [hunk 15]
```diff
@@ -527,7 +528,6 @@ def assign(self, build: _T, host: _T) -> None:
     def __repr__(self) -> str:
         return f'PerMachine({self.build!r}, {self.host!r})'
 
-
```
Comment (eli-schwartz) [inline mesonbuild/utils/universal.py:530]: Unrelated change?

File: mesonbuild/compilers/rust.py [hunk 16]
```diff
@@ -33,6 +33,35 @@
     's': ['-C', 'opt-level=s'],
 }
 
+def get_rustup_run_and_args(exelist: T.List[str]) -> T.Optional[T.Tuple[T.List[str], T.List[str]]]:
+    """Given the command for a rustc executable, check if it is invoked via
+       "rustup run" and if so separate the "rustup [OPTIONS] run TOOLCHAIN"
+       part from the arguments to rustc.  If the returned value is not None,
+       other tools (for example clippy-driver or rustdoc) can be run by placing
+       the name of the tool between the two elemnts of the tuple."""
```
Comment (eli-schwartz) [inline mesonbuild/compilers/rust.py:41]:
```suggestion
       the name of the tool between the two elements of the tuple."""
```

File: mesonbuild/compilers/rust.py [hunk 17]
```diff
@@ -245,6 +275,26 @@ def get_assert_args(self, disable: bool, env: 'Environment') -> T.List[str]:
         action = "no" if disable else "yes"
         return ['-C', f'debug-assertions={action}', '-C', 'overflow-checks=no']
 
+    def get_rust_tool(self, name: str, env: 'Environment') -> T.List[str]:
```
Comment (eli-schwartz) [inline mesonbuild/compilers/rust.py:278]: No need to quote annotations since we use `from __future__ import annotations`. It's not really worth the churn of rewriting tons of older code (that can be done when modifying the code anyway) but when writing new code please omit the quotes.

File: mesonbuild/scripts/run_tool.py [hunk 18]
```diff
@@ -127,3 +129,10 @@ def run_clang_tool(name: str, srcdir: Path, builddir: Path, fn: T.Callable[...,
     def wrapper(path: Path) -> T.Iterable[T.Coroutine[None, None, int]]:
         yield fn(path, *args)
     return asyncio.run(_run_workers(all_clike_files(name, srcdir, builddir), wrapper))
+
+def run_tool_on_targets(fn: T.Callable[[T.Dict[str, T.Any]],
+                                       T.Iterable[T.Coroutine[T.Any, T.Any, int]]]) -> int:
+    targets = load_info_file(get_infodir(os.getcwd()), kind='targets')
```
Comment (eli-schwartz) [inline mesonbuild/scripts/run_tool.py:135]: I don't think we need to import mintro and its attendant module dependency tree *just* so we can programmatically get filenames which are documented as a stable API in https://mesonbuild.com/IDE-integration.html

File: docs/markdown/howtox.md [hunk 19]
```diff
@@ -239,6 +239,26 @@ And then pass it through the variable (remember to use absolute path):
...
+```console
+$ meson setup builddir
+$ ninja -C builddir clippy
+```
+
+Clippy will also obey the `werror` [builtin option](Builtin-options.md#core-options).
+
+By default Meson uses as many concurrent processes as there are cores
+on the test machine. You can override this with the environment
+variable `MESON_TESTTHREADS`.
```
Comment (eli-schwartz) [inline docs/markdown/howtox.md:255]: Pursuant to my previous comment: documenting the use of MESON_TESTTHREADS in this context doesn't make this a clean naming scheme.

File: mesonbuild/backend/ninjabackend.py [hunk 20]
```diff
@@ -3639,6 +3644,20 @@ def generate_dist(self) -> None:
...
+        if 'clippy' in self.all_outputs or not self.have_language('rust'):
+            return
+
+        cmd = self.environment.get_build_command() + \
+            ['--internal', 'clippy', self.environment.build_dir]
+        elem = self.create_phony_target('clippy', 'CUSTOM_COMMAND', 'PHONY')
+        elem.add_item('COMMAND', cmd)
+        elem.add_item('pool', 'console')
+        for crate in self.rust_crates.values():
+            if crate.crate_type in {"rlib", "dylib", "proc-macro"}:
```
Comment (eli-schwartz) [inline mesonbuild/backend/ninjabackend.py:3657]: Inconsistent quotation style.

**Review by eli-schwartz — CHANGES_REQUESTED** [review]

File: mesonbuild/scripts/run_tool.py (hunk 10 above)
Comment (bonzini) [reply mesonbuild/scripts/run_tool.py:7]: I find it clearer to import both but it's the same. :)

File: mesonbuild/scripts/run_tool.py (hunk 11 above)
Comment (bonzini) [reply mesonbuild/scripts/run_tool.py:28]: In mtest it's there because it also includes the environment settings, which is not really possible to write without Unix syntax afaik. Will use join_args here.

File: mesonbuild/scripts/run_tool.py (hunk 12 above)
Comment (bonzini) [reply mesonbuild/scripts/run_tool.py:34]: Hmm yeah, it's like ninja in this respect but it's unfortunate for clang-tidy. I will make it respect b_colorout=always/auto.

File: mesonbuild/scripts/run_tool.py (hunk 13 above)
Comment (bonzini) [reply mesonbuild/scripts/run_tool.py:56]: It is clearly useful for more than `meson test` though. I will add support for MESON_THREADS, with extra fallback to MESON_TESTTHREADS for `meson test` only.

File: mesonbuild/scripts/run_tool.py (hunk 14 above)
Comment (bonzini) [reply mesonbuild/scripts/run_tool.py:81]: Or moved to utils, but I think a one-off is okay since these are the two places that use asyncio in Meson.

File: mesonbuild/scripts/run_tool.py (hunk 12 above)
Comment (eli-schwartz) [reply mesonbuild/scripts/run_tool.py:34]: Since this uses the console pool, ninja directly connects run_tool consumers such as `ninja clippy clang-format clang-tidy` to the tty, and os.isatty(sys.stdout) returns true. `ninja clang-format-check` etc. will actually produce colored statuses (albeit potentially interleaved a bit randomly).

The question is how to use b_colorout. Would that mean just to disable the asyncio pool and disable stdout pipe buffering? Is it possible to conveniently fake `isatty()` the way e.g. /usr/bin/script does?

File: mesonbuild/scripts/run_tool.py (hunk 11 above)
Comment (bonzini) [reply mesonbuild/scripts/run_tool.py:28]: quoted_cmdline is only used for printing on stdout btw.

File: mesonbuild/scripts/run_tool.py (hunk 12 above)
Comment (bonzini) [reply mesonbuild/scripts/run_tool.py:32]: Yeah, I meant "it's like ninja" in the sense that for compilers you need --color=always due to ninja's buffering (which is what I'm trying to mimic here).

> The question is how to use b_colorout

Since we know what tools we're running, we can add --color=always|never options by hand in mesonbuild/scripts. For clang-tidy that's actually a bugfix in the opposite case: `ninja clang-tidy` was not respecting -Db_colorout=never.

For clippy and (in the future) rustfmt we get the `--color=always|auto|never option` for free from the rustc command line.

> Is it possible to conveniently fake isatty() the way e.g. /usr/bin/script does?

script uses a pty which is pretty hard to do, especially for Windows. We can fake it by using the isatty value inside the script when -Db_colorout=auto, i.e. using isatty to change auto to `always` or `never`.

Comment (eli-schwartz) [reply mesonbuild/scripts/run_tool.py:32]: Basically what I mean to say here is that I have some kind of recollection that I tried to fix interleaved output here a couple years ago and threw up my hands in disgust upon learning that clang tooling didn't actually support command line options to enable color anyways.

Maybe it's changed but on the other hand it looks like your latest push only handles tidy?

Comment (bonzini) [reply mesonbuild/scripts/run_tool.py:32]: yeah, but I found out clang-format also supports -f{,no-}color-diagnostics so it can be fixed as well

Comment (eli-schwartz) [reply mesonbuild/scripts/run_tool.py:32]: Could be informative to check which version of clang first added the undocumented support.

Comment (bonzini) [reply mesonbuild/scripts/run_tool.py:32]: If it's since [https://github.com/llvm/llvm-project/commit/692175399453d70083772033e13a11e408576327](this) it would be LLVM 7.0.0. I'll try to find some old distros, but it may not be this week.

**Comment by bonzini** [comment]
Ok, I checked and indeed clang-format's -fcolor-diagnostics option simply does not work, while the replacement works as back as clang-format 10 (the one that introduced `--dry-run -Werror`). I documented the fun debugging experience of LLVM/clang's overengineered C++ in the commit message.

**Review by dcbaker — APPROVED** [review]

File: mesonbuild/utils/universal.py [hunk 21]
```diff
@@ -1086,6 +1088,29 @@ def default_sysconfdir() -> str:
     return 'etc'
 
 
+def determine_worker_count(varnames: T.Optional[T.List[str]] = None) -> int:
+    num_workers = 0
+    varnames = varnames or []
+    varnames.append('MESON_NUM_PROCESSES')
+    for varname in varnames:
+        if varname in os.environ:
```
Comment (eli-schwartz) [inline mesonbuild/utils/universal.py:1096]: So the idea here is to use the last varname that exists?

Can we document the order used here?

File: mesonbuild/scripts/clippy.py [hunk 22]
```diff
@@ -0,0 +1,66 @@
...
+class ClippyDriver:
+    def __init__(self, build: build.Build, tempdir: str):
+        self.tools: PerMachine[T.List[str]] = PerMachine([], [])
+        self.warned: T.DefaultDict[str, bool] = defaultdict(lambda: False)
+        self.tempdir = tempdir
+        for machine in MachineChoice:
+            compilers = build.environment.coredata.compilers[machine]
+            if 'rust' in compilers:
+                compiler = compilers['rust']
+                assert isinstance(compiler, RustCompiler)
```
Comment (eli-schwartz) [inline mesonbuild/scripts/clippy.py:24]: This assert is the only reason we're importing compilers.rust, but the reason for the assert seems to be solely for type checking.

The assert pulls in a runtime dependency but allows using python -OO to optimize it out to nothing. The cost/benefit ratio here isn't great. We can use `T.cast('RustCompiler', compilers['rust'])` instead, which allows moving the import to T.TYPE_CHECKING

**Review by eli-schwartz — CHANGES_REQUESTED** [review]
Overall this seems good based on an eyeball check. A couple minor nits though, and I haven't tested it myself so I'll defer merging to another maintainer.

Thanks for the excellent clang-format research work. :)

File: mesonbuild/scripts/clippy.py [hunk 23]
```diff
@@ -0,0 +1,67 @@
...
+                compiler = T.cast('RustCompiler', compilers['rust'])
+                self.tools[machine] = compiler.get_rust_tool('clippy-driver', build.environment)
+
+    def warn_missing_clippy(self, machine: str) -> None:
+        if self.warned[machine]:
+            return
+        mlog.warning(f'clippy-driver not found for {machine} machine')
+        self.warned[machine] = True
+
+    def __call__(self, target: T.Dict[str, T.Any]) -> T.Iterable[T.Coroutine[T.Any, T.Any, int]]:
```
Comment (dcbaker) [inline mesonbuild/scripts/clippy.py:34]: Pretty sure that Coroutine is like Generator, so `[None, None, int]`

**Review by dcbaker — COMMENTED** [review]
One small thing that I think (but I'm not sure) can be improved in the typing, otherwise I'm still fine with this. If using `None` doesn't work feel free to dismiss.

File: mesonbuild/scripts/clippy.py (hunk 23 above)
Comment (bonzini) [reply mesonbuild/scripts/clippy.py]: Yep, that works (I fixed it in one of the commits but not this one).

File: docs/markdown/IDE-integration.md [hunk 24]
```diff
@@ -93,13 +93,21 @@ can provide code completion for all source files.
 ```json
 {
     "language": "language ID",
+    "machine": "build" / "host",
```
Comment (dnicolodi) [inline docs/markdown/IDE-integration.md:96]: At a quick look, the other parts of this JSON snippet are valid JSON, while the one added here is not. I don't know how relevant this is but it may be confusing to have invalid example data. Unfortunately JSON does not allow comments, thus I don't really know how to fix this without dropping the information that `machine` is an enumerated type.
